src/goto-synthesizer/cegis_verifier.cpp

```diff
@@ -0,0 +1,524 @@
+        goto_model.get_goto_function(from_fun_name);
+      for(goto_programt::const_targett it =
+            from_function.body.instructions.begin();
+          it != from_function.body.instructions.end();
+          ++it)
+      {
+        if(it->location_number == step.pc->location_number)
+        {
+          from = it;
+        }
+      }
+
+      // Get `to`---the instruction where the violation happens
+      irep_idt to_fun_name = goto_trace.get_last_step().function_id;
+      const goto_functionst::goto_functiont &to_function =
+        goto_model.get_goto_function(to_fun_name);
+      for(goto_programt::const_targett it =
+            to_function.body.instructions.begin();
+          it != to_function.body.instructions.end();
+          ++it)
+      {
+        if(it->location_number == violation->location_number)
+        {
+          to = it;
+        }
+      }
+
+      // The violation is caused by the loop havoc
+      // if it is dependent on the loop havoc.
+      if(dependence_graph.is_flow_dependent(from, to))
+      {
+        // TODO: not safe for inlined static functions.
+        result.function_id = step.pc->source_location().get_function();
+        result.loop_number = step.pc->loop_number;
+      }
+    }
+  }
+  return result;
+}
+
+cext cegis_verifiert::build_cex(
+  const goto_tracet &goto_trace,
+  const source_locationt &loop_entry_loc)
+{
+  const namespacet ns(goto_model.symbol_table);
+
+  // Valuations of havoced variables right after havoc instructions.
+  std::map<exprt, std::size_t> object_sizes;
+  std::map<exprt, std::size_t> havoced_values;
+  std::map<exprt, std::size_t> havoced_pointer_offsets;
+
+  // Loop-entry valuations of havoced variables.
+  std::map<exprt, std::size_t> loop_entry_values;
+  std::map<exprt, std::size_t> loop_entry_offsets;
+
+  // Live variables upon the loop head.
+  std::set<exprt> live_variables;
+
+  bool entered_loop = false;
+
+  // Scan the trace step by step to store needed valuations.
+  for(const auto &step : goto_trace.steps)
+  {
+    switch(step.type)
+    {
+    case goto_trace_stept::typet::DECL:
+    case goto_trace_stept::typet::ASSIGNMENT:
+    {
+      if(!step.full_lhs_value.is_nil())
+      {
+        // Entered loop?
+        if(is_assignment_to_instrumented_variable(step.pc, ENTERED_LOOP))
+          entered_loop = step.full_lhs_value == true_exprt();
+
+        // skip hidden steps
+        if(step.hidden)
+          break;
+
+        // Live variables
+        // 1. must be in the same function as the target loop;
+        // 2. alive before entering the target loop;
+        // 3. a pointer or a primitive-typed variable;
+        // TODO: add support for union pointer
+        if(
+          step.pc->source_location().get_function() ==
+            loop_entry_loc.get_function() &&
+          !entered_loop &&
+          (step.full_lhs.type().id() == ID_unsignedbv ||
+           step.full_lhs.type().id() == ID_pointer) &&
```

Comment:
They are the type of variables we enumerate in the synthesized invariants. That is, we only enumerate expressions containing pointer types and unsigned primitive types in the current implementation.
I wouldn't say this subset is the right one. This subset is the least one we need to synthesize loop invariants for aws-c-common and coreJSON benchmarks. Because memory safety is about pointer, size, and offset, I would say this set is a reasonable set of types we consider when we are synthesizing loop invariants for memory safety. 
Maybe `signedbv` should also be included.

Reply:
+1 to adding `signedbv` in here.